source/extensions/common/async_files/async_file_manager_thread_pool.cc

```diff
@@ -0,0 +1,253 @@
+#include "source/extensions/common/async_files/async_file_manager_thread_pool.h"
+
+#include <memory>
+#include <queue>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include "source/extensions/common/async_files/async_file_action.h"
+#include "source/extensions/common/async_files/async_file_context_thread_pool.h"
+#include "source/extensions/common/async_files/status_after_file_error.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace AsyncFiles {
+
+namespace {
+// ThreadNextAction is per worker thread; if enqueue is called from a callback
+// the action goes directly into ThreadNextAction, otherwise it goes into the
+// queue and is eventually pulled out into ThreadNextAction by a worker thread.
+thread_local std::shared_ptr<AsyncFileAction> ThreadNextAction;
+
+// ThreadIsWorker is set to true for worker threads, and will be false
+// for all other threads.
+thread_local bool ThreadIsWorker = false;
+} // namespace
+
+AsyncFileManagerThreadPool::AsyncFileManagerThreadPool(
+    const envoy::extensions::common::async_files::v3::AsyncFileManagerConfig& config,
+    Api::OsSysCalls& posix)
+    : posix_(posix) {
+  if (!posix.supportsAllPosixFileOperations()) {
+    throw EnvoyException("AsyncFileManagerThreadPool not supported");
+  }
+  unsigned int thread_pool_size = config.thread_pool().thread_count();
+  if (thread_pool_size == 0) {
+    thread_pool_size = std::thread::hardware_concurrency();
+  }
+  ENVOY_LOG(info, fmt::format("AsyncFileManagerThreadPool created with id '{}', with {} threads",
+                              config.id(), thread_pool_size));
+  thread_pool_.reserve(thread_pool_size);
+  while (thread_pool_.size() < thread_pool_size) {
+    thread_pool_.emplace_back([this]() { worker(); });
+  }
+}
+
+AsyncFileManagerThreadPool::~AsyncFileManagerThreadPool() ABSL_LOCKS_EXCLUDED(queue_mutex_) {
+  {
+    absl::MutexLock lock(&queue_mutex_);
+    terminate_ = true;
+  }
+  while (!thread_pool_.empty()) {
+    thread_pool_.back().join();
+    thread_pool_.pop_back();
+  }
+}
+
+std::string AsyncFileManagerThreadPool::describe() const {
+  return absl::StrCat("thread_pool_size = ", thread_pool_.size());
+}
+
+std::function<void()> AsyncFileManagerThreadPool::enqueue(std::shared_ptr<AsyncFileAction> action) {
+  auto cancelFunc = [action]() { action->cancel(); };
+  absl::MutexLock lock(&queue_mutex_);
+  if (ThreadIsWorker) {
```

Comment:
Can you add a comment here? This is an optimization if this is being called from within a callback running on a worker thread? Technically couldn't you avoid grabbing the lock in this case and just acquire the lock below?

Reply:
Good catch on the lock saving. It's not just an optimization, it's an intentional behavior difference, the new action doesn't yield the thread, essentially skipping to the head of the queue - added a comment to that effect.